xieyu_wang/generator.cpp

```diff
@@ -0,0 +1,130 @@
+#include "generator.h"
+Generator::Generator(int maxDepth, int newVerticesNum)
+    : maxDepth_(maxDepth), newVerticesNum_(newVerticesNum) {
+  if (maxDepth < 0) {
+    while (maxDepth_ < 0) {
+      std::cout << "Error Max Depth must be  above or equal 0: ";
+      std::cin >> maxDepth_;
+    }
+  }
+  if (newVerticesNum < 0) {
+    while (newVerticesNum_ < 0) {
+      std::cout << "Error new Vertices number must be above or equal 0: ";
+      std::cin >> newVerticesNum_;
+    }
+  }
+}
+
+bool Generator::vertxGenerateProbabilityController(int depth) {
+  float prob = (maxDepth_ - depth);
+  prob = prob / maxDepth_;
+  std::bernoulli_distribution probability(prob);
+  return probability(generator_);
+}
+
+Graph Generator::generateGrey(int presentDepth = 1, int fromVertexId = 0) {
+  Graph graph;
+  graph.addVertex();
+  std::vector<int> verticesPresentLevel;
+  for (int depth = 0; depth < maxDepth_; depth++) {
+    const auto& vertexIds = graph.getVertexIdsAtDepth(depth);
+    bool isNewVertexGenerated = false;
+    for (const auto& vertexId : vertexIds) {
+      for (int i = 0; i < newVerticesNum_; ++i) {
+        if (vertxGenerateProbabilityController(presentDepth)) {
+          const auto& newVertex = graph.addVertex();
+          graph.addEdge(vertexId, newVertex.id, Edge::Color::Gray);
+          isNewVertexGenerated = true;
+        }
+      }
+    }
+    if (!isNewVertexGenerated) {
+      break;
+    }
+  }
+  return graph;
+}
+void Generator::generateGreen(Graph& graph) {
+  int verticesSum = graph.getPresentVertexId();
+  for (int i = 0; i < verticesSum; ++i) {
+    if (specialEdgeGenerateProbabilityController(Edge::Color::Green)) {
+      graph.addEdge(i, i, Edge::Color::Green);
+    }
+  }
```

Comment:
I think I told you, that you don't need this function `getPresentVertexId()`
```suggestion
  for (auto& vertex : graph.vertices()) {
    if (is_lucky(GREEN_EDGE_PROBABILITY)) {
      graph.addEdge(vertex.id, vertex.id, Edge::Color::Green);
    }
  }
```